File: easy/third_max.py
import sys

sentinel = -sys.maxsize - 1


def third_max(nums: list[int]) -> int:
    unique = set(nums)
    first, second, third = sentinel, sentinel, sentinel
    for n in unique:
        if first < n:
            first, second, third = n, first, second
        elif second < n:
            second, third = n, second
        elif third < n:
            third = n
    if len(unique) < 3:
        return first
    return third

File: easy/third_max.go
package easy

import (
	"maps"
	"math"
)

func ThirdMax(nums []int) int {
	unique := make(map[int]bool)
	for _, n := range nums {
		unique[n] = true
	}
	first, second, third := math.MinInt, math.MinInt, math.MinInt
	for n := range maps.Keys(unique) {
		switch {
		case n > first:
			third = second
			second = first
			first = n
		case n > second:
			third = second
			second = n
		case n > third:
			third = n
		}
	}
	if len(unique) < 3 {
		return first
	}
	return third
}
